js/app/downloader.js

```diff
@@ -204,51 +169,38 @@ const SubscriptionDownloader = new Lang.Class({
-        // Verify that client shards actually appear on disk.
-        old_shards = old_shards.filter((shard) => {
-            let shard_file = directory.get_child(shard.path);
-            return shard_file.query_exists(cancellable);
-        });
+    _download_new_shards_promise: function (directory, old_manifest, new_manifest, cancellable) {
+        return Promise.resolve()
+        .then(() => {
+            let old_shards = old_manifest.shards;
+            // Verify that client shards actually appear on disk.
+            old_shards = old_shards.filter((shard) => {
+                let shard_file = directory.get_child(shard.path);
+                return shard_file.query_exists(cancellable);
+            });
 
-        let new_shards = new_manifest.shards;
+            let new_shards = new_manifest.shards;
 
-        // Download any shards that are not on the client.
+            // Download any shards that are not on the client.
 
-        // XXX: We're currently indexing by filename, assuming it will be good enough.
-        // Will it be, or should we also store e.g. ETag?
-        let to_download = subtract_set(new_shards, old_shards, (shard) => shard.path);
-        let to_delete = subtract_set(old_shards, new_shards, (shard) => shard.path);
+            // XXX: We're currently indexing by filename, assuming it will be good enough.
+            // Will it be, or should we also store e.g. ETag?
+            let to_download = subtract_set(new_shards, old_shards, (shard) => shard.path);
+            let to_delete = subtract_set(old_shards, new_shards, (shard) => shard.path);
 
-        if (to_download.length === 0 && to_delete.length === 0) {
-            task.return_value(false);
-            return task;
-        }
+            if (to_download.length === 0 && to_delete.length === 0)
+                return false;
 
-        AsyncTask.all(this, (add_task) => {
-            to_download.forEach((shard) => {
+            let downloads = to_download.map((shard) => {
                 let shard_file = directory.get_child(shard.path);
                 let shard_uri = Soup.URI.new(shard.download_uri);
                 let shard_csum = shard.sha256_csum;
 
                 let request = new FileDownloadRequest(shard_uri, shard_file, shard_csum);
-                add_task((cancellable, callback) => this._file_downloader.download_file(request, cancellable, callback),
-                         (task) => this._file_downloader.download_file_finish(task));
+                return this._file_downloader.download_file_promise(request, cancellable);
             });
-        }, cancellable, task.catch_callback_errors((source, download_shards_task) => {
-            AsyncTask.all_finish(download_shards_task);
-
-            task.return_value(true);
-        }));
-
-        return task;
-    },
-
-    _download_new_shards_finish: function (task) {
-        return task.finish();
+            return Promise.all(downloads).then(function () { return true; });
```

Comment:
I think I was trying to keep the old API, where the true would be returned if an update was actually applied, false if no update was applied, and throwing if an error occurred.

Though I need to make sure to actually return that value in `fetch_update_promise` ill fix that up.